### DISCUSSION TITLE
Lesson 4: unable to execute print(transaction)

### DISCUSSION THREAD

**Discussion by ti51120** [smartcontractkit/full-blockchain-solidity-course-py#407] [Q&A] [answered]
Can anyone help with the error below? It works perfectly fine until I try to execute the print(transaction). 

![image](https://user-images.githubusercontent.com/52118731/142709968-8eebd921-3505-4461-a94f-999f0de4bc57.png)

Below is my code.

```python
from solcx import compile_standard, install_solc
import json
from web3 import Web3

with open("./SimpleStorage.sol", "r") as file:
    simpel_storage_file = file.read()

# Compile solidity
install_solc("0.6.0")
compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simpel_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {"*": ["abi", "metadata", "evm.bytecode", "evm.sourceMap"]}
            }
        },
    }
)

with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)

bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]

#  Connecting to ganache
w3 = Web3(Web3.HTTPProvider("HTTP://127.0.0.1:7545"))
chain_id = 1337
my_address = "0x27e0A545B03bFE673f036C108df49465e68D6F65"
private_key = "0x15530a613dd53507c305eeba6f5e0995bf356d6fbbd0a9a7e882357163f20c6d"

# Create the contract in Python
SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)
nonce = w3.eth.getTransactionCount(my_address)

transaction = SimpleStorage.constructor().buildTransaction(
    {"chainId": chain_id, "from": my_address, "nonce": nonce}
)

print(transaction)
```

**Comment by cromewar** [comment, marked as answer]
Hello @ti51120 that's strange, let me understand the error just show up when adding the printing statement? I read the error and the `maxPriorityFeeGas` is usually related with Alchemy. @PatrickAlphaC could you give us a hand on this one?

**Reply by ti51120** [reply]
Hello @cromewar,
Today I try to install web3 with version 5.24.0 and it works. 

![image](https://user-images.githubusercontent.com/52118731/142723359-47fc66b8-ec85-4224-af6d-0245025cb392.png)

Although I still don't know why, it seems that
 ```python 
transaction = SimpleStorage.constructor().buildTransaction(
    {"chainId": chain_id, "from": my_address, "nonce": nonce}
)
```
this piece of code won't work for web3-5.25.0.

![image](https://user-images.githubusercontent.com/52118731/142723343-2722af62-6975-42e2-903f-3974709c56a8.png)

**Reply by cromewar** [reply]
I'm almost sure that it has something to be with the python packages communicating with each other, and in some way that version has conflicts with other packages. Managing python packages on windows is honestly a mess, I strongly suggest you to use python virtual environments for each project and also if you want a challenge try using WSL.

Anyways as your problem seems to be solved I'll close this thread please feel free to make any other questions.

**Reply by PatrickAlphaC** [reply]
Nice work :)

**Reply by Zetazero25** [reply]
I am running into the same issue as discussed in this thread. It seems that you are suggesting i change which version of web3 i have installed. How do i know which version of web3 is installed? and if the wrong version is installed how do i install the correct version?

**Reply by cromewar** [reply]
Hello @Zetazero25 you can see the web3 versions [here](https://pypi.org/project/web3/#history).
If you want to install a specific version of a package on python using pip just do this: `pip install package==version`.

**Reply by PatrickAlphaC** [reply]
https://stackoverflow.com/questions/70104101/valueerror-method-eth-maxpriorityfeepergas-not-supported-web3-py-with-ganache

**Reply by Zetazero25** [reply]
Thank You! you guys are amazing

**Comment by yujiosword** [comment]
For visability, the answer can be found here: https://stackoverflow.com/questions/70104101/valueerror-method-eth-maxpriorityfeepergas-not-supported-web3-py-with-ganache

Encountered same error before, and it can be resolved by adding one more argument when building a transaction
```python
transaction = SimpleStorage.constructor().buildTransaction(
    {
        "gasPrice": w3.eth.gas_price,
        "chainId": chain_id,
        "from": my_address,
        "nonce": nonce,
    }
)
```

**Reply by cromewar** [reply]
You are awesome @yujiosword .
